sinatra メモアプリ

yama998877 opened the pull request:
メモのタイトルと内容を保存、変更、削除できます。

yama998877 committed:
sinatra メモアプリ 提出

yama998877 committed:
Update README.md

タイトルがおかしかったので編集

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,106 @@
+  File.open(filename, 'w') do |file|
+    file.puts(word)
+  end
+  # File.open(filename).close
```
不要な行であれば削除しましょう。

yama998877 replied:
すみません。削除しました。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,106 @@
+  end
+end
+
+get '/' do
```
> RESTに倣ったURIになっていること

という要件がありますので、これを満たすようにPath設計を見直してみてください。

yama998877 replied:
見直して修正しました🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,106 @@
+  save_memo("./data/#{SecureRandom.uuid}.json", memo_hash.to_json)
+  @names = read_memo
+  @memo_path = Dir.children('./data/').sort_by { |f| File.mtime("./data/#{f}") }
+  redirect '/'
+  erb :index
```
`redirect` しているのに `erb` 指定は不要かと思うのですが、どうでしょうか。

yama998877 replied:
不要でした。erb指定は削除しました。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,106 @@
+end
+
+delete '/' do
+  params
```
これは何のための行でしょうか？

yama998877 replied:
デバックに使ったのを消し忘れていました。申し訳ありません。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,106 @@
+  memo_hash = {}
+  @title = params['title']
+  @title << 'タイトルがありません' if @title.strip.empty?
+  memo_hash['title'] = h(params['title'])
+  memo_hash['detail'] = h(params['detail'])
```
HTMLエスケープ処理は基本、view側で行います。
なぜかというと、元のテキストが失われてしまうので、記録的な意味でそのままとっておきたいのが一点、
また、例えば、CSVなどHTML以外のフォーマットに出力したい、という要望が今後追加されたとすると、HTMLエスケープ処理は必要ないですよね。ということで、「HTMLエスケープしないといけない」というのはview側の都合なんですよね。なので、データ自体は改変せずに、view側で都度エスケープして表示する、というのが自然かなと思います！
SinatraのFAQにも[こんな感じのサンプルコード](https://sinatrarb.com/faq.html#escape_html)があります

yama998877 replied:
アドバイスありがとうございます！
erbファイルで表示するときにのみHTMLエスケープ処理をするように変更して、元のテキストが失われないようにしました！

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,106 @@
+post '/' do
+  memo_hash = {}
+  @title = params['title']
+  @title << 'タイトルがありません' if @title.strip.empty?
```
ユーザーの入力データを変更してしまうのはちょっと考えたほうがよいです。
ユーザーからすると自分が入力したはずのデータが勝手に書き換えられるのは直感的ではないですし、また、今後このタイトルがない場合の表示を `(タイトル未入力)` にしたい、という要件がもしでたときにデータを全部変更しないといけなくなります。
タイトルがないことを明示したい場合は `view` 側でやるようにし、データ自体はそのまま保存するか、エラー表示をしてタイトルがない状態で保存できないようにしたほうがよいかと思います。

yama998877 replied:
アドバイスありがとうございます！
erbファイルで表示するときにのみ`タイトルがありません`に変更して、データ自体はそのまま保存するようにしました。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,106 @@
+  json_files = []
+  @memo_path = Dir.glob('./data/*.json').sort_by { |f| File.mtime(f) }
+  @memo_path.each do |file|
+    p json_files << File.read(file)
```
`p` しているのはなぜでしょうか？デバッグ用であれば提出物からは消してしまってください。

yama998877 replied:
申し訳ありません。消去しました。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,106 @@
+
+delete '/' do
+  params
+  delete_file = params['json']
+  File.delete("./data/#{delete_file}")
```
これはセキュリティ上、かなり問題があります。
たとえば `params['json']` として `../app.rb` が渡されたらどうなるでしょうか？
このコード自体を削除してしまいます。
このプロセスを実行するユーザーの権限によってば、任意のファイル削除が可能です。
データの保存方式を見直すか、必ず `data` 以下のファイルしか触れないようにチェックしてください。

yama998877 replied:
削除機能を実装することばかりに気を取られていて気づいていませんでした。
`./data`以下のファイルしか消せないようにしました。
ありがとうございます🙏

yoshitsugu reviewed (commented):
不要そうな行が残っていたり、要件が満たされていなかったり、ちょっと提出物が雑な印象を受けました。もうちょっとご自身で品質を高める工夫を考えてみるとよいかなと思います。

yama998877 committed:
app.rbの修正

不要なコード、コメントの削除
URLの改善
HTMLエスケープ処理機能の修正
./data以下のファイル以外のファイルを消せないようにif文を追加

yama998877 committed:
detail.erbの修正

app.rbで行っていたHTMLエスケープ処理をview側で表示するときだけ適用するように変更
URLを修正

yama998877 committed:
index.erbの修正

ブロック変数名が不適切に思えたので修正
<a>のname属性を削除
メモにタイトルがなかった場合の処理を追加

yama998877 committed:
layout.rbの修正

リンク先のURLを修正

yama998877 committed:
new_datail.erbの修正

form actionのURLを変更

yama998877 committed:
new.erb

form actionのURLを修正

yama998877 committed:
Merge branch 'sinatra-practice' of https://github.com/yama998877/practice-memoapp into sinatra-practice
リモートリポジトリのREADME.mdをgithubで編集したため

yoshitsugu commented on views/new_detail.erb:
```diff
@@ -0,0 +1,15 @@
+  <input type="hidden" name="_method" value="PATCH">
+  <input type='hidden' name='json' value='<%= @file_name %>'>
+
+  タイトル:<input type="text" name="title" value="<%= @memo['title'] %>">
```
ここもエスケープが必要です。たとえば `"<script>alert("test")</script>` など `"` つきの文字がはいってしまうとscriptを実行されてしまいます。

yama998877 replied:
ご指摘ありがとうございます。修正しました🙇‍♂️

yoshitsugu commented on views/new_detail.erb:
```diff
@@ -0,0 +1,15 @@
+
+  <div>
+  メモ:
+  <textarea name="detail" rows="5"><%= @memo['detail'] %></textarea>
```
こちらも `</textarea><script>alert("test")</script>` などとすると実行できますね。
こういう「終端できてしまう」ものは脆弱性につながるので、十分注意してください。

yama998877 replied:
ご指摘ありがとうございます。こちらもエスケープ処理をしました🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+  @memo_path = Dir.glob('./data/*.json').sort_by { |f| File.mtime(f) }
+  @memo_path.each do |file|
+    json_files << File.read(file)
+  end
```
`read_memo` という読み込み用のメソッドがあるのに、ここでそれが使われてないのがちょっともったいないですね。共通化できないか考えてみてください。

yama998877 replied:
ご指摘ありがとうございます。共通化できました。ありがとうございます！

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+  hashs = []
+  json_files.each do |json|
+    hashs << JSON.parse(json)
+  end
```
```ruby
array = []
original_array.each do |item|
  array << item.xxx
end
```

の形のものは

```ruby
array = original_array.map { |item| item.xxx }
```

のように `map` を使うと簡単に書けます。
配列から別の配列を作るときは `map` を使う、と覚えておくとよいかなーと思います〜
https://docs.ruby-lang.org/ja/latest/method/Array/i/collect.html

yoshitsugu replied:
また `hashs` だと何をあらわしているのかわからないので、ちゃんとわかるような命名にしてください。
特にRubyのような型で情報を明示しない言語において、変数名は重要です。
Rubyの作者であるMatzも[名前重要](https://プログラマが知るべき97のこと.com/%E3%82%A8%E3%83%83%E3%82%BB%E3%82%A4/%E5%90%8D%E5%89%8D%E9%87%8D%E8%A6%81/)というコラムをかいていたりします。

yama998877 replied:
ご指摘ありがとうございます。`map`を使うように変更しました！
変数名も変更しました！

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+  @titles = []
+  hashs.each do |t|
+    @titles << t['title']
+  end
```
こちらも同様ですね。

yama998877 replied:
こちらも修正しました！

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+  end
+end
+
+def read_memo
```
配列を返す場合は `read_memos` のように複数形にするとわかりやすいです。

yama998877 replied:
修正しました。ありがとうございます！

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+  memo_hash['title'] = params['title']
+  memo_hash['detail'] = params['detail']
+  save_memo("./data/#{SecureRandom.uuid}.json", memo_hash.to_json)
+  @names = read_memo
```
これは何に使われていますか？

yama998877 replied:
動作確認用のコードです。消し忘れていました。申し訳ありません。削除しました。🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+
+post '/memos' do
+  memo_hash = {}
+  @title = params['title']
```
この `@title`  も使われていなさそうですね。

yama998877 replied:
これも削除し忘れです。本当に申し訳ありません。🙇

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+  memo_hash['detail'] = params['detail']
+  save_memo("./data/#{SecureRandom.uuid}.json", memo_hash.to_json)
+  @names = read_memo
+  @memo_path = Dir.children('./data/').sort_by { |f| File.mtime("./data/#{f}") }
```
これも使われてなさそう。

yama998877 replied:
本当に申し訳ありません。🙇
削除しました。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+patch '/memos' do
+  new_memo_hash = {}
+  new_memo_hash['title'] = params['title']
+  new_memo_hash['detail'] = params['detail']
```
あまりHashだから、というのをあらわすのに `xxx_hash` という命名は使わないですね。
(Rubyはduck typingですし。。)
また、
```ruby
new_memo = {
  'title' => params['title'],
  'detail' => params['detail']
}
```
のようにHashリテラルを使うほうが見やすいかなと思います。

yama998877 replied:
確かにhashリテラルの方がhashだとわかりやすいですし、見やすいですね！
修正しました。ご指摘ありがとうございます。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+  new_memo_hash = {}
+  new_memo_hash['title'] = params['title']
+  new_memo_hash['detail'] = params['detail']
+  update_file = params['json']
```
この行だけ見ると 「`update_file` が `params['json']` とはどういうことだ？」とわからなくなってしまうので、
たとえば

```ruby
file_name = "./data/#{params['memo_id']}.json"
```

のようにしたほうがよいのではと思いました。

yama998877 replied:
修正しました。ありがとうございます！

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+  redirect '/memos'
+end
+
+patch '/memos' do
```
これはREST準拠ではないですね。

yama998877 replied:
ご指摘ありがとうございます。修正しました。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+  redirect '/memos'
+end
+
+delete '/memos' do
```
ここもREST準拠ではなさそうです！

yama998877 replied:
ここも修正しました。ご指摘ありがとうございます！

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+end
+
+delete '/memos' do
+  delete_file = params['json']
```
ここも上記と同様です。`json` はJSON形式のことなので、その他の意味に使わないようにしましょう。

yama998877 replied:
手直しした結果、このコードは使わなくなりましたがご指摘ありがとうございます。`json`のみでは使わないように修正しました。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+  erb :new
+end
+
+get '/memos/data/:file' do
```
これもREST準拠になっていません。
RESTのプラクティスを再度復習してみてください。

yama998877 replied:
ご指摘ありがとうございます。修正しました。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+  @file_name = params[:file]
+  @memo = JSON.parse(File.read("./data/#{@file_name}"))
+  erb :new_detail
+end
```
`new_detail` が編集ページであることはなかなか想起しづらいので、 `edit` とかにしましょう。
RESTはAPIを目的としているので、新規作成フォーム、編集フォームは標準的なPathはないのですが、RESTらしい `/リソース名の複数形` もしくは `/リソース名の複数形/リソースID` の形に近いものにならないか考えてみてください。

yama998877 replied:
editの方がいいですね！ありがとうございます。修正しました。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+delete '/memos' do
+  delete_file = params['json']
+  delete_path = Pathname.new("./data/#{delete_file}")
+  File.delete(delete_path) if delete_path.dirname.to_s == './data'
```
もうちょっと自由度を狭めたほうがよいかなと思います。
リクエストで渡せるのは `uuid` のみとし、
```ruby
# 最初に存在するuuidかをチェックする
uuids = Dir.glob('./data/*.json').map { |filename| File.basename(filename, ".json") }
# 存在する場合だけ削除
if uuids.include?(uuid)
  File.delete('"./data/#{uuid}.json")
end
```
のようにすればより安全です。

yama998877 replied:
ありがとうございます。修正しました。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,100 @@
+get '/memos/data/:file' do
+  @json_file = params[:file]
+
+  @memo_detail = JSON.parse(File.read("./data/#{@json_file}"))
```
ここも関係ないファイルの中身を表示されてしまう可能性があります。
あらかじめ配列に存在するファイルをとっておき、そこから検索するなど、安全になるようにしてください。

yama998877 replied:
修正しました。ありがとうございます。

yoshitsugu reviewed (commented):
コメントしました。

yama998877 committed:
new_detail.erbをedit.erbに変更

yama998877 committed:
修正

変更ボタンの移動先のURLを変更
formのname属性の変数名を変更

yama998877 committed:
リンクのURLの修正

 ルーティングのURLを変更してページが表示されなくなったのでmemos/jsonファイル名になるように修正

yama998877 committed:
修正

メソッド名を変更
変数名を変更

yama998877 committed:
変数、ルーティングの順番の整理、不要なコメントの削除

hashが含まれていた変数の名前を変更。URL名を変更したので順番を見やすいように入れ替え。

yama998877 committed:
gemの追加

追加する前の状態だとbundle exec ruby app.rbで起動できなくなったので追加

yama998877 committed:
form actionのURL変更

ルーティングのURLを変更したのでこちらも変更。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,88 @@
+  erb :new
+end
+
+get '/memos/:file' do
```
特にこの後 `json` ファイルではなくDBへの保存のプラクティスもありますし、Pathにjsonという情報がはいってしまうのはあまりよくないかなと思います。idだけPathに指定すればいいようにしましょう。

yama998877 replied:
idだけpathに指定できるようにしました。
ありがとうございます🙇‍♂️

yama998877 replied:
申し訳ありません。前回のやりとりのみで意図がよくわからないのですがidのみpathに指定できるようにしたのに`:file`がルーティングあるのは良くないということでしょうか？
`:file`を`:id`に変更しました。勘違いでしたら申し訳ありません🙇

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,88 @@
+    'title' => params['title'],
+    'detail' => params['detail']
+  }
+  edited_file_name = "./data/#{params['memo_id']}"
```
Pathの引数からファイル名はわかるはずなのに、なぜここで別のparamsを参照しているのでしょうか？

yama998877 replied:
前回のコメントで

> この行だけ見ると 「update_file が params['json'] とはどういうことだ？」とわからなくなってしまうので、
> たとえば
> 
> ```ruby
> file_name = "./data/#{params['memo_id']}.json"
> ```
> のようにしたほうがよいのではと思いました。

とコメントしてもらったので修正しました。

`patch '/memos'`を`patch '/memos/:file'`にしたので不要でした。気づかず修正してしまい申し訳ありません🙇
削除してpathの引数を使うようにしました。ご指摘ありがとうございます🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,88 @@
+  ascending_order_files.each do |file|
+    memo_files << File.read(file)
+  end
+end
```
全体的に、

- ファイルからのメモ一覧の読み出し: `load_memos`
- ファイルからの特定のメモの読み出し: `load_memo`
- ファイルへの書き出し: `save_memos`

として、同様の処理をしている箇所はまとめませんか？
あちこちに同じような処理をしているコードがちらばっていて、無駄なように思いました。

`read` -> `load` としたのは `save` との対称性をとるためです。逆に読み込みを `read` とするなら書き込みは `write` のほうがいいですね。

yoshitsugu replied:
ちょっと苦戦している感じなので、コードのヒントも書いておくと、

```ruby
DATA_DIR = File.join(File.dirname(__FILE__), 'data')

def load_memos
  Dir.glob("#{DATA_DIR}/*.json").sort_by { |f| File.mtime(f) }.map do |f| 
    memo = # ファイル名からJSONにまで変換しておく
    memo[:uuid] = # ファイル名からuuidを抽出しておく
    memo
  end
end
```

のようにすると `uuid` を持ったmemo一覧ができるので、 `load_memo` のほうは `load_memos` した結果を `uuid` で絞り込むだけで済みますね。

yama998877 replied:
コードのヒントありがとうございます！参考にしてメソッドを作り直しました！

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,88 @@
+def save_memo(filename, word)
+  File.open(filename, 'w') do |file|
+    file.puts(word)
+  end
```
これは `filename` に渡された任意のファイルに書き出せてしまうのでセキュリティ的にかなり問題ですね。
`../app.rb` を渡すようにするとこのRubyファイル自体を書き換えたりできそうです。

yama998877 replied:
ご指摘ありがとうございます。
メソッド内に`return if filename.include?('..')`を追加して`..`がpathに含まれていたら実行されないようにしました。

yoshitsugu reviewed (commented):
コメントしました！
詰まったら早めに相談するようにおねがいします〜

yama998877 committed:
ファイルに書き込むメソッドに条件を追加

他のディレクトリに書き込まれたり、他のディレクトリのファイルに書き込まれないようにファイルパスに`..`があったらreturnするようにした

yama998877 committed:
メモの読み出しメソッドの修正、書き出しメソッド名の変更、ルーティング内の処理の修正、定数の追加、params['memo_id']の削除。

read_memoメソッドの追加、save_memoメソッドをwrite_memoに名前変更、read_memosメソッド内のコードを修正、定数DATA_DIRを追加、メソッドを修正して影響の出たルーティングの処理を修正。

yama998877 committed:
memo_idは不要になったので削除

yama998877 committed:
新しい呼び出しメソッドでtitleとファイルのuuidの順番が変わったので変更。

yama998877 commented:
> コメントしました！ 詰まったら早めに相談するようにおねがいします〜

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,90 @@
+end
+
+def read_memo(memo_uuid)
+  read_memos.detect { |memo_data| memo_data.value?(memo_uuid) }
```
これだと本文に該当のuuidがあってもマッチしてしまいませんか？ちゃんと `memo[:uuid]` のvalueとして該当のものがあるかどうかをチェックしましょう。

yama998877 replied:
ご指摘ありがとうございます。`memo[:uuid]`のvalueと`memo_uuid`が一致するときのみ実行されるように条件を追加しました🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,90 @@
+end
+
+get '/memos' do
+  @url_titles = read_memos.map { |memo_url| memo_url.values_at('title', :uuid) }
```
`'title', :uuid` と混在しているのはややこしいので、keyはsymbolに統一したほうがいいのかなと思います。

yama998877 replied:
ご指摘ありがとうございます。symbolに統一しました🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,90 @@
+  }
+  edited_file_name = "./data/#{params[:file]}.json"
+  changing_content = edited_memo.to_json
+  write_memo(edited_file_name, changing_content)
```
これだと `params[:file]` としていろいろなファイル名をわたせば任意のファイル名を指定できてしまいませんか？そうするとここの本来の意味の更新だけではなく追加もできてしまいそうです。

yama998877 replied:
ご指摘ありがとうございます。すでにファイルがあるかread_memoで調べてnilだったら実行されないように条件を追加しました🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,90 @@
+
+get '/memos/:file/edit' do
+  @file_name = params[:file]
+  @memo = JSON.parse(File.read("./data/#{@file_name}.json"))
```
ここも `read_memo` が使えるのではないでしょうか。

yama998877 replied:
ご指摘ありがとうございます。使うように修正しました🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,90 @@
+
+delete '/memos/:file' do
+  uuid = params[:file]
+  uuids = read_memos.map { |memo| memo[:uuid] }
+  File.delete("./data/#{uuid}.json") if uuids.include?(uuid)
```
判定方法としてこれでもよいのですが、せっかく `read_memo` があるので、 `read_memo` してみて `nil` じゃなければ削除、としたほうが簡単に思います。

yama998877 replied:
ご指摘ありがとうございます。read_memoを使うように修正しました🙇‍♂️

yoshitsugu commented on views/index.erb:
```diff
@@ -0,0 +1,16 @@
+    <% @url_titles.each do |title,link| %>
+    <li><a href="memos/<%= File.basename("#{link}") %>">
+    <%=
+    if title.strip.empty?
+      "タイトルがありません"
```
親切心からだと思いますが、これだと 「タイトルがありません」というタイトルを入力したようにも見えますし、何か表示をわかりやすくする(括弧の中に入れるとか文字を灰色にするなど)か、そもそもタイトル入力を必須にしましょう。

yama998877 replied:
ご指摘ありがとうございます。「」の中に入れて文字色を灰色にしました🙇‍♂️

yoshitsugu reviewed (commented):
コメントしました！

yama998877 committed:
app.rbの修正

ハッシュのkeyを文字列からsymbolになるように変更。
read_memoメソッドに:uuidの値がuuidの時、条件が一致するか確認するコードを追加。
ルーティングのパスの`:file`を`:id`に変更。
メモの変更時に新しいメモのファイルが追加できないように条件を追加。
削除の時の条件にread_memoをしようするように変更。
@memoをread_memoをしようするように変更。

yama998877 committed:
警告されたので修正

ifではなくunlessを使ってくださいとRuboCopに言われたので修正。

yama998877 committed:
erbファイルの修正

タイトルが入力されなかった場合に表示されるタイトルを「タイトルがありません」に変更。さらに灰色に文字色を変更。
detail.erbの不要なコードの削除。
edit.erbとdetail.erbのkeyを文字列からsymbolに変更

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,89 @@
+def read_memos
+  Dir.glob("#{DATA_DIR}/*.json").sort_by { |f| File.mtime(f) }.map do |f|
+    memo = JSON.parse(File.read(f), symbolize_names: true)
+    memo[:uuid] = f.gsub(%r{./data/|.json}, '')
```
- 複数あるので、 `memos` と複数形のほうがいいと思います。
- `File.basename(f, '.json')` で大丈夫だと思います。
- `memos` をHashとして、 `memos[uuid] = memo` の形式にすればread_memo側も `memos[uuid]` とすればよく簡単かと思います。

yama998877 replied:
ご指摘ありがとうございます🙏
修正しました🙇

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,89 @@
+end
+
+def write_memo(filename, word)
+  return if filename.include?('..')
```
`unless read_memo(uuid).nil?` されていることを前提とするなら、ここのチェックは不要と思います。そうでないなら、同様のチェックをここでしたほうがよいです。
`..` だけのチェックだとその他の場合はないか？などの懸念があります。こういうのはAllowListとDenyListという考え方があります。セキュリティ面では基本拒否(Deny)でAllowListを運用するほうがセキュアになりやすいです。興味あれば調べてみてください。

yama998877 replied:
> unless read_memo(uuid).nil? されていることを前提とするなら、ここのチェックは不要と思います。そうでないなら、同様のチェックをここでしたほうがよいです。

削除して動作確認してみたところ、必要ありませんでした。
ご指摘ありがとうございます🙏

yoshitsugu reviewed (commented):
コメントしました。

yama998877 committed:
read_memosメソッドの修正、それによって影響のでた箇所の修正、write_memoメソッドのreturn if 削除

read_memosメソッドを{uuid =>{title: hoge, detail: hoge}になるように変更。
read_memoメソッド内のread_memosメソッドが使えるように修正。
`get '/memos'`内の変数`@url_titles`を削除。
index.erbの`@url_titles`を`read_memos`に修正。
ブロック変数をlinkとmemoに変更。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,88 @@
+  read_memos[memo_uuid]
+end
+
+def write_memo(filename, word)
+  File.open(filename, 'w') do |file|
+    file.puts(word)
+  end
+end
```
`word` だと単語という意味なので、あまりどういう引数なのか判然としません。
`read_memos` と同じ形の `memos` 、もしくは変更対象の `memo` のHashをうけとって、その後の保存処理はメソッド内に隠蔽してくれると使う側からすると使いやすいかなと思います。

yama998877 replied:
ご指摘ありがとうございます🙏
`write_memo(uuid, memo_title, memo_detail)`にしてメソッド内でHashに処理、保存するようにしました🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,88 @@
+    title: params[:title],
+    detail: params[:detail]
+  }
+  edited_file_name = "./data/#{uuid}.json"
+  changing_content = edited_memo.to_json
```
上にも書いたとおり、この辺は保存時にのみ必要になる処理だと思うので、 `write_memos`に隠蔽できるといいですね。

yama998877 replied:
ご指摘ありがとうございます🙏
write_memo内で処理するように修正しました🙇‍♂️

yoshitsugu reviewed (commented):
コメントしました！

yama998877 committed:
post,patchのルーティング内の修正、メソッドの修正

post,patchのルーティング内のメモの保存時に必要になる処理をwrite_memoにまとめた。

yoshitsugu commented on app.rb:
```diff
@@ -0,0 +1,83 @@
+  memos
+end
+
+def read_memo(memo_uuid)
+  read_memos[memo_uuid]
+end
```
これあまりメソッドがある意味がないですね。 `read_memos[memo_uuid]` とそのまま書いちゃったほうがわかりやすいように思いました。

yama998877 replied:
ご指摘ありがとうございます🙇‍♂️
削除して該当箇所を修正しました。

yoshitsugu reviewed (commented):
一点コメントしました！

yama998877 committed:
read_memoメソッドの削除

無いほうがわかりやすいので削除。
read_memoを使用していた所にはread_memosを代わりに使う。

yoshitsugu reviewed (approved):
OKです！
